diff --git conf/inter_athena.conf conf/inter_athena.conf
index 13498ebc0..d89f176bf 100644
--- conf/inter_athena.conf
+++ conf/inter_athena.conf
@@ -29,7 +29,7 @@ party_share_level: 15
 login_server_ip: 192.168.1.95
 login_server_port: 3306
 login_server_id: root
-login_server_pw: Gdhk8hk[
+login_server_pw: gdhk8hk[
 login_server_db: Ragnarok2
 login_codepage:
 login_case_sensitive: no
@@ -37,7 +37,7 @@ login_case_sensitive: no
 ipban_db_ip: 192.168.1.95
 ipban_db_port: 3306
 ipban_db_id: root
-ipban_db_pw: Gdhk8hk
+ipban_db_pw: gdhk8hk[
 ipban_db_db: Ragnarok2
 ipban_codepage:
 
@@ -45,21 +45,21 @@ ipban_codepage:
 char_server_ip: 192.168.1.95
 char_server_port: 3306
 char_server_id: root
-char_server_pw: Gdhk8hk
+char_server_pw: gdhk8hk[
 char_server_db: Ragnarok2
 
 // MySQL Map Server
 map_server_ip: 192.168.1.95
 map_server_port: 3306
 map_server_id: root
-map_server_pw: Gdhk8hk
+map_server_pw: gdhk8hk[
 map_server_db: Ragnarok2
 
 // MySQL Log Database
 log_db_ip: 192.168.1.95
 log_db_port: 3306
 log_db_id: root
-log_db_pw: Gdhk8hk
+log_db_pw: gdhk8hk[
 log_db_db: Ragnarok2
 log_codepage:
 log_login_db: loginlog
